crates/noirc_evaluator/src/ssa/code_gen.rs

```diff
@@ -0,0 +1,1006 @@
+        HirForExpression, HirLiteral,
+    },
+    stmt::{HirConstrainStatement, HirLetStatement, HirPrivateStatement, HirStatement},
+};
+use noirc_frontend::node_interner::{ExprId, FuncId, IdentId, StmtId};
+//use noirc_frontend::{FunctionKind, Type};
+use num_bigint::BigUint;
+
+// This is a 'master' class for generating the SSA IR from the AST
+// It contains all the data; the node objects representing the source code in the nodes arena
+// and The CFG in the blocks arena
+// everything else just reference objects from these two arena using their index.
+pub struct IRGenerator<'a> {
+    pub context: Option<&'a Context>,
+
+    pub first_block: arena::Index,
+    pub current_block: arena::Index,
+    pub blocks: arena::Arena<block::BasicBlock>,
+    pub nodes: arena::Arena<node::NodeObj>,
+    pub id0: arena::Index, //dummy index.. should we put a dummy object somewhere?
+    pub value_name: HashMap<arena::Index, u32>,
+    pub sealed_blocks: HashSet<arena::Index>,
+}
+
+impl<'a> IRGenerator<'a> {
+    pub fn new(context: &Context) -> IRGenerator {
+        let mut pc = IRGenerator {
+            context: Some(context),
+            id0: IRGenerator::dummy_id(),
+            first_block: IRGenerator::dummy_id(),
+            current_block: IRGenerator::dummy_id(),
+            blocks: arena::Arena::new(),
+            nodes: arena::Arena::new(),
+            // dummy_instruction: ParsingContext::dummy_id(),
+            value_name: HashMap::new(),
+            sealed_blocks: HashSet::new(),
+        }; //, objects: arena::Arena::new()
+        block::create_first_block(&mut pc);
+        pc
+    }
+
+    fn print_operand(&self, idx: arena::Index) -> String {
+        let var = self.get_object(idx);
+        if var.is_none() {
+            return format!("unknown {:?}", idx);
+        }
+        match var.unwrap() {
+            node::NodeObj::Obj(v) => v.print(),
+            node::NodeObj::Instr(i) => i.print_i(),
+            node::NodeObj::Const(c) => c.print(),
+        }
+    }
+
+    pub fn print_block(&self, b: &block::BasicBlock) {
+        for idx in &b.instructions {
+            let ins = self.get_instruction(*idx);
+            let mut str_res;
+            if ins.res_name.is_empty() {
+                str_res = format!("{:?}", idx);
+            } else {
+                str_res = ins.res_name.clone();
+            }
+            if ins.is_deleted {
+                str_res += " -DELETED";
+            }
+            let lhs_str = self.print_operand(ins.lhs);
+            let rhs_str = self.print_operand(ins.rhs);
+            let ins_str = format!(
+                "{} {} {}",
+                lhs_str,
+                format!(" op:{:?} ", ins.operator),
+                rhs_str
+            );
+            println!("{}: {}", str_res, ins_str);
+
+            if ins.operator == node::Operation::phi {
+                //    dbg!(&ins.phi_arguments);
```

Comment:
Should phi args be printed as well?

Reply:
Yes, but should be prettier than the dbg!(..).

Reply:
Done